utils/src/main/kotlin/com/exactpro/th2/lwdataprovider/MessageSearcher.kt

```diff
@@ -0,0 +1,191 @@
+ */
+class MessageSearcher private constructor(
+    private val service: DataProviderService,
+    private val searchStep: Duration = DEFAULT_SEARCH_STEP,
+    private val defaultBook: String,
+    private val defaultSessionGroup: String,
+    private val defaultMessageStreams: Set<MessageStream>,
+) {
+    /**
+     * Searches the first message in [BASE64_FORMAT] format matched by filter. Execute search from [Instant.now] to previous time
+     * @param sessionGroup is parameter for request. Searcher uses [messageStreams] value when [sessionGroup] is null or blank
+     * @param searchInterval for searching. Searcher stops to send requests when the next `from` time is before than search start time minus [searchInterval]
+     */
+    @JvmOverloads
+    fun findLastOrNull(
+        searchInterval: Duration,
+        book: String = defaultBook,
+        sessionGroup: String = defaultSessionGroup,
+        messageStreams: Set<MessageStream> = defaultMessageStreams,
+        filter: Function<MessageSearchResponse, Boolean>
+    ): MessageSearchResponse? {
+        require(book.isNotBlank()) {
+            "'book' can't be blank"
+        }
+        require(sessionGroup.isNotBlank()) {
+            "'sessionGroup' can't be blank"
+        }
+        messageStreams.forEach {
+            require(it.name.isNotBlank()) {
+                "'name' in ${it.toJson()} message stream can't be blank"
+            }
+            require(it.direction != UNRECOGNIZED) {
+                "'direction' in ${it.toJson()} message stream can't be $UNRECOGNIZED"
+            }
+        }
+
+        require(!searchInterval.isZero) {
+            "'searchInterval' can't be zero"
+        }
+        require(!searchInterval.isNegative) {
+            "'searchInterval' can't be negative"
+        }
+
+        val now = Instant.now()
+        val end = now.minus(searchInterval)
+        var from: Instant
+        var to = now
+
+        val requestBuilder = createRequestBuilder(
+            book = book,
+            sessionGroup = sessionGroup,
+            messageStreams = messageStreams
+        )
+        K_LOGGER.info { "Search the last or null, base request: ${requestBuilder.toJson()}, start time: $now, end time: $end" }
+
+        return withCancellation {
+            generateSequence {
+                from = to
+                to = from.minus(searchStep).run { if (this < end) end else this }
+                if (from == to) { return@generateSequence null }
+
+                service.searchMessageGroups(
+                    requestBuilder.apply {
+                        startTimestamp = from.toTimestamp()
+                        endTimestamp = to.toTimestamp()
+                    }.build().also {
+                        K_LOGGER.debug { "Request from: $from, to: $to" }
+                    }
+                )
+            }.flatMap(Iterator<MessageSearchResponse>::asSequence)
```

Comment:
I would rewrite this with `sequence` method instead. This will allow us to keep all the variables in one place instead of capturing a mutable reference to them in the lambda that we use as a supplier for the next element. But you can leave it as is if you want

Reply:
Could you please add a few more details for your suggestion

Reply:
There is a `sequence` method in Kotlin that provides you with a sequence builder based on Kotlin continuation (like in coroutines). Using this method you can write something like:
```kotlin
val seq: Sequence<Int> = sequence {
  var i = 0
  while (i < 50) {
    yield(i)
    i += 1
  }
}
```
This helps to hold every variable inside and avoid using mutable references like in your case (`from` can not be mutated from inside the lambda). You will get something like this in the bytecode
```java
final Ref.ObjectRef from = new Ref.ObjectRef();
from.element = Instant.now();
```